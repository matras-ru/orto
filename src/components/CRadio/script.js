import { selfInstall } from '@/utils';
import commonAttributes from '@/mixins/commonAttributes.js';

import ThemeClass from '@/themes/default/CRadio';

const {
    disabledClass,
    baseIconClass,
    defaultIconClass,
    checkedIconClass,
    disabledIconClass,
    labelClass,
    inputClass,
    wrapperClass
} = ThemeClass;

export default {
    name: 'Radio',

    install(Vue, theme) {
        selfInstall(Vue, theme, this);
    },

    props: {
        ...commonAttributes.props,
        label: {
            type: String,
            default: null
        },
        modelValue: {
            type: [String, Number, Boolean],
            default: null
        },
        value: {
            type: [String, Number, Boolean],
            default: null
        }
    },

    model: {
        prop: 'modelValue',
        event: 'change'
    },

    data() {
        return {
            isChecked: false
        };
    },

    computed: {
        currentClass() {
            let classes = [labelClass];

            if (this.disabled) {
                classes.push(disabledClass);
            }

            return classes;
        },
        currentIconClass() {
            let classes = [baseIconClass];

            if (this.disabled) {
                classes.push(disabledIconClass);
            } else if (this.isChecked || this.shouldBeChecked) {
                classes.push(checkedIconClass);
            } else {
                classes.push(defaultIconClass);
            }

            return classes;
        },
        shouldBeChecked() {
            return this.modelValue === this.value;
        }
    },

    methods: {
        onChange(e) {
            this.$emit('change', e.target.checked ? this.value : null);
        }
    },

    render(h) {
        const attrs = {
            id: this.id,
            autofocus: this.autofocus,
            name: this.name,
            type: 'radio',
            disabled: this.disabled
        };

        const domProps = {
            checked: this.shouldBeChecked,
            value: this.value
        };

        const on = {
            change: this.onChange
        };

        const componentData = {
            class: inputClass,
            attrs,
            domProps,
            on
        };

        return h('label', { class: this.currentClass, attrs: { for: this.id } }, [
            h('input', componentData),
            h('div', { class: wrapperClass }, [
                h('span', { class: this.currentIconClass }),
                h('span', this.label)
            ])
        ]);
    }
};
