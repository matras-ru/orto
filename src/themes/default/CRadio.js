const Radio = {
    disabledClass: 'text-tertiary-200',
    baseIconClass: 'w-0-8 h-0-8 border-2 mr-0-5 mt-0-2 rounded-full',
    defaultIconClass: 'border-black-200 bg-white',
    checkedIconClass: 'border-black-200 bg-secondary-200 shadow-inner',
    disabledIconClass: 'border-tertiary-200 bg-white'
};

export default Radio;
