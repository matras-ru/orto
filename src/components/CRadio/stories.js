import { storiesOf } from '@storybook/vue';
import { action } from '@storybook/addon-actions';
import { withKnobs, text, boolean } from '@storybook/addon-knobs';

import CRadio from './';

const stories = storiesOf('CRadio', module);
stories.addDecorator(withKnobs);

stories
    .add('Radio knobs', () => ({
        components: { CRadio },
        props: {
            id: {
                default: text('id', 'radio1')
            },
            autofocus: {
                default: boolean('autofocus', false)
            },
            label: {
                default: text('label', 'Radio1')
            },
            name: {
                default: text('name', 'radio1')
            },
            disabled: {
                default: boolean('disabled', false)
            }
        },
        template:
            '<CRadio @onChange="action" :id="id" :autofocus="autofocus" :name="name" :label="label" :disabled="disabled" />',
        methods: { action: action('change') }
    }))
    .add('Default radio', () => ({
        components: { CRadio },
        template: '<CRadio @onChanget="action" id="radio2" label="Radio2" name="radio2" />',
        methods: { action: action('change') }
    }))
    .add('Disabled radio', () => ({
        components: { CRadio },
        template: '<CRadio @onChange="action" id="radio3" label="Radio3" name="radio3" disabled />',
        methods: { action: action('change') }
    }));
