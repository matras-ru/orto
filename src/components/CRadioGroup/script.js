import { selfInstall } from '@/utils';
import CRadio from '@/components/CRadio';

export default {
    name: 'RadioGroup',

    install(Vue, theme) {
        selfInstall(Vue, theme, this);
    },

    components: {
        CRadio
    },

    props: {
        data: {
            type: Array,
            default: () => []
        },
        modelValue: {
            type: [String, Number, Boolean],
            default: null
        }
    },

    model: {
        prop: 'modelValue',
        event: 'change'
    },

    computed: {
        value: {
            get() {
                return this.modelValue;
            },
            set(val) {
                this.$emit('change', val);
            }
        }
    },

    render(h) {
        let setProps = item => {
            return {
                id: item.id,
                autofocus: this.autofocus,
                label: item.label,
                name: item.name,
                disabled: item.disabled,
                value: item.value
            };
        };

        return h(
            'div',
            this.data.map(item => {
                return h('CRadio', {
                    props: {
                        modelValue: this.value,
                        ...setProps(item)
                    },
                    on: {
                        change: val => {
                            this.value = val;
                        }
                    }
                });
            })
        );
    }
};
