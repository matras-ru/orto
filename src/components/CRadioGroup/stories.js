import { storiesOf } from '@storybook/vue';

import CRadioGroup from './';

const stories = storiesOf('CRadioGroup', module);

stories.add('CRadioGroup', () => ({
    components: { CRadioGroup },
    template:
        "<CRadioGroup :data=\"[{ id: 'radio2', label: 'Radio1', name: 'radio2' }, { id: 'radio3', label: 'Radio2', name: 'radio3' }]\" />"
}));
