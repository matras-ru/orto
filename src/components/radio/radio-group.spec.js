/* eslint-env jest */
import { mount, config, createLocalVue } from '@vue/test-utils';

import CRadioGroup from './radio-group';
import CRadio from './radio';

config.stubs['CRadio'] = CRadio;

describe('Radio group', () => {
    it('is rendered & not empty', () => {
        const wrapper = mount(CRadioGroup, {
            context: {
                props: {
                    modelValue: null,
                    data: [
                        {
                            id: 'radio1',
                            label: 'radio1',
                            name: 'radio1',
                            value: 'radio1'
                        },
                        {
                            id: 'radio2',
                            label: 'radio2',
                            name: 'radio2',
                            value: 'radio2'
                        }
                    ]
                }
            }
        });

        expect(wrapper.findAll('input[type="radio"]').length).toBe(2);
    });

    it('v-model', () => {
        const localVue = new createLocalVue();

        const App = localVue.extend({
            data() {
                return {
                    model: null
                };
            },

            render(h) {
                return h(CRadioGroup, {
                    props: {
                        data: [
                            {
                                id: 'radio1',
                                label: 'radio1',
                                name: 'radio1',
                                value: 'radio1'
                            },
                            {
                                id: 'radio2',
                                label: 'radio2',
                                name: 'radio2',
                                value: 'radio2'
                            }
                        ],
                        modelValue: this.model
                    },
                    on: {
                        change: val => {
                            this.model = val;
                        }
                    }
                });
            }
        });

        const wrapper = mount(App, {
            localVue: localVue
        });

        const input = wrapper.find('input[type="radio"]');

        expect(wrapper.vm.model).toEqual(null);

        input.element.selected = true;
        input.trigger('click');

        expect(wrapper.vm.model).toEqual('radio1');

        wrapper.destroy();
    });
});
