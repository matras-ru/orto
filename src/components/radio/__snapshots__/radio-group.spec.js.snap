// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio group is rendered & not empty 1`] = `
<div class="mb-2-2">
  <div>
    <div>
      <div class="flex flex-wrap mb-0-4"><label for="radio1" class="relative pl-1-4 cursor-pointer"><input id="radio1" name="radio1" type="radio" class="absolute top-0-2 left-0 w-0-8 h-0-8 form-radio" value="radio1">radio1</label></div>
      <div class="flex flex-wrap mb-0-4"><label for="radio2" class="relative pl-1-4 cursor-pointer"><input id="radio2" name="radio2" type="radio" class="absolute top-0-2 left-0 w-0-8 h-0-8 form-radio" value="radio2">radio2</label></div>
    </div>
  </div>
</div>
`;
